### PR TITLE
Refactor l4 tests

### PR DESCRIPTION
**Pull request by panslava**
- Split l4netlb_test assertions by functions
- Move common l4_test, l4netlb_test assertions to l4testutils

This will also make it easier to add dual-stack tests to l4netlb in the future

/assign @kl52752

### CODE REVIEW

File: pkg/loadbalancers/l4_test.go [hunk 1]
```diff
@@ -628,11 +628,11 @@ func TestEnsureInternalLoadBalancerDeletedWithSharedHC(t *testing.T) {
 	nodeNames := []string{"test-node-1"}
 	namer := namer_util.NewL4Namer(kubeSystemUID, nil)
 
-	_, _, result := ensureService(fakeGCE, namer, nodeNames, vals.ZoneName, 8080, t)
+	_, _, result := ensureService(fakeGCE, namer, nodeNames, vals.ZoneName, t)
 	if result != nil && result.Error != nil {
 		t.Fatalf("Error ensuring service err: %v", result.Error)
 	}
-	svc2, l4, result := ensureService(fakeGCE, namer, nodeNames, vals.ZoneName, 8081, t)
+	svc2, l4, result := ensureService(fakeGCE, namer, nodeNames, vals.ZoneName, t)
```
Comment (kl52752) [inline pkg/loadbalancers/l4_test.go:635]: Is it working as intended? Do we expect here 2 different services or ensure twice the same service?

File: pkg/loadbalancers/l4_test.go [hunk 2]
```diff
@@ -704,7 +704,7 @@ func TestHealthCheckFirewallDeletionWithNetLB(t *testing.T) {
 	}
 }
 
-func ensureService(fakeGCE *gce.Cloud, namer *namer_util.L4Namer, nodeNames []string, zoneName string, port int, t *testing.T) (*v1.Service, *L4, *L4ILBSyncResult) {
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4_test.go:707]: I agree, this is much better solution :) fixed

File: pkg/loadbalancers/l4netlb_test.go [hunk 3]
```diff
@@ -219,125 +218,195 @@ func ensureLoadBalancer(port int, vals gce.TestClusterValues, fakeGCE *gce.Cloud
...
-		t.Errorf("Unexpected protocol '%s' for backend service %v", bs.Protocol, bs)
+		return nil, fmt.Errorf("unexpected protocol '%s' for backend service %v", bs.Protocol, bs)
 	}
+	backendServiceLink := cloud.SelfLink(meta.VersionGA, l4netlb.cloud.ProjectID(), "backendServices", key)
 	if bs.SelfLink != backendServiceLink {
-		t.Errorf("Unexpected self link in backend service - Expected %s, Got %s", bs.SelfLink, backendServiceLink)
-	}
-	if bs.LoadBalancingScheme != string(cloud.SchemeExternal) {
-		t.Errorf("Unexpected load balancing scheme - Expected EXTERNAL, Got %s", bs.LoadBalancingScheme)
+		return nil, fmt.Errorf("unexpected self link in backend service - Expected %s, Got %s", bs.SelfLink, backendServiceLink)
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb_test.go]: Thank you, done

File: pkg/loadbalancers/l4netlb_test.go [hunk 4]
```diff
@@ -219,125 +218,195 @@ func ensureLoadBalancer(port int, vals gce.TestClusterValues, fakeGCE *gce.Cloud
...
+}
+
+func verifyNetLBForwardingRule(l4netlb *L4NetLB, backendServiceLink string) error {
+	frName := l4netlb.GetFRName()
+	fwdRule, err := composite.GetForwardingRule(l4netlb.cloud, meta.RegionalKey(frName, l4netlb.cloud.Region()), meta.VersionGA)
 	if err != nil {
-		t.Errorf("Failed to fetch forwarding rule %s - err %v", frName, err)
+		return fmt.Errorf("failed to fetch forwarding rule %s - err %w", frName, err)
 	}
 	if fwdRule.Name != frName {
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb_test.go]: I agree, wanted to delete it, but then thought, that I want to refactor, but not change the logic that much. But I will remove it

File: pkg/loadbalancers/l4netlb_test.go [hunk 5]
```diff
@@ -219,125 +218,195 @@ func ensureLoadBalancer(port int, vals gce.TestClusterValues, fakeGCE *gce.Cloud
...
+		return fmt.Errorf("unexpected name for forwarding rule '%s' - expected '%s'", fwdRule.Name, frName)
 	}
+	if fwdRule.LoadBalancingScheme != string(cloud.SchemeExternal) {
+		return fmt.Errorf("unexpected LoadBalancingScheme for forwarding rule '%s' - expected '%s'", fwdRule.LoadBalancingScheme, cloud.SchemeExternal)
+	}
+
+	proto := utils.GetProtocol(l4netlb.Service.Spec.Ports)
 	if fwdRule.IPProtocol != string(proto) {
-		t.Errorf("Unexpected protocol '%s' for forwarding rule %v", fwdRule.IPProtocol, fwdRule)
+		return fmt.Errorf("unexpected protocol '%s' for forwarding rule %v", fwdRule.IPProtocol, fwdRule)
```
[the remark this answers is not in the document]
Comment (panslava) [reply pkg/loadbalancers/l4netlb_test.go]: Thank you, done

**Pull request merged**